src/moonlink/src/storage/wal.rs

```diff
@@ -1 +1,331 @@
+/// Note that wal manager is meant to be used in a single thread. While
+/// persist and delete_files can be called concurrently, their results
+/// have to be handled serially.
+pub(crate) struct WalManager {
+    in_mem_wal: InMemWal,
+    /// The wal file numbers that are still live.
+    live_wal_file_tracker: Vec<WalFileInfo>,
+    /// Tracks the file number to be assigned to the next persisted wal file
+    curr_file_number: u64,
+
+    file_system_accessor: Arc<dyn BaseFileSystemAccess>,
+}
+
+impl WalManager {
+    pub fn new(config: FileSystemConfig) -> Self {
+        // TODO(Paul): Add a more robust constructor when implementing recovery
+        Self {
+            in_mem_wal: InMemWal::new(0),
+            live_wal_file_tracker: Vec::new(),
+            curr_file_number: 0,
+            file_system_accessor: Arc::new(FileSystemAccessor::new(config)),
+        }
+    }
+
+    pub fn insert(&mut self, table_event: &TableEvent) {
+        let wal_event = WalEvent::new(table_event, self.in_mem_wal.highest_lsn);
+        self.in_mem_wal.buf.push(wal_event);
+
+        // Update highest_lsn if this event has a higher LSN
+        if let Some(lsn) = table_event.get_lsn_for_ingest_event() {
+            assert!(lsn >= self.in_mem_wal.highest_lsn);
+            self.in_mem_wal.highest_lsn = lsn;
+        }
+
+        // TODO(Paul): Implement streaming flush (if cross threshold, begin streaming write)
+    }
+
+    fn take(&mut self) -> Vec<WalEvent> {
+        std::mem::take(&mut self.in_mem_wal.buf)
+    }
+
+    pub fn get_file_name(file_number: u64) -> String {
+        format!("wal_{file_number}.json")
+    }
+
+    pub fn get_to_persist_wal_file_info(&self) -> WalFileInfo {
+        WalFileInfo {
+            file_number: self.curr_file_number,
+            highest_lsn: self.in_mem_wal.highest_lsn,
+        }
+    }
+
+    pub async fn persist(
+        file_system_accessor: Arc<dyn BaseFileSystemAccess>,
+        wal_to_persist: &Vec<WalEvent>,
+        wal_file_info: &WalFileInfo,
+    ) -> Result<()> {
+        if !wal_to_persist.is_empty() {
+            let wal_json = serde_json::to_vec(&wal_to_persist).unwrap();
+
+            let wal_file_path = WalManager::get_file_name(wal_file_info.file_number);
+            file_system_accessor
+                .write_object(&wal_file_path, wal_json)
+                .await?;
+        }
+        Ok(())
+    }
+
+    fn handle_complete_persist(&mut self, persist_and_truncate_result: &PersistAndTruncateResult) {
+        if let Some(file_persisted) = &persist_and_truncate_result.file_persisted {
+            assert!(file_persisted.file_number == self.curr_file_number);
+            // Add the persisted file to the live tracker
+            self.live_wal_file_tracker.push(file_persisted.clone());
+            self.curr_file_number += 1;
+        }
+    }
+
+    pub fn get_files_to_truncate(&self, truncate_from_lsn: u64) -> Vec<WalFileInfo> {
```

Comment:
nit: this function took me some time to read through, do you mind adding a short comment?
(not merge blocker)